### ISSUE TITLE
Renewal date does not update

### ISSUE THREAD

**Issue by tannerm** [pippinsplugins/restrict-content-pro#279]
When I subscript without auto renewing and then update to a subscription and select auto-renew, the renewal date does not update to the new subscription.

**Comment by pippinsplugins** [comment]
@tannerm Are you referring to when editing a member's subscription in the admin?

**Comment by tannerm** [comment]
No, this is when a member chooses a membership without auto-renewing it and then switches to a different membership with auto renew.

**Comment by pippinsplugins** [comment]
Does the expiration date get set properly when subscribing to a recurring subscription from the get go?

It sounds like you don't have webhooks / IPN configured or working properly.

**Comment by tannerm** [comment]
ah... that makes sense. I saw this while developing locally. I'll test on a staging site.

**Comment by pippinsplugins** [comment]
Some gateways update the expiration date during a webhook call. Stripe (for example) does.

**Comment by kaizau** [comment]
@tannerm I'm running into the same problem. You may need to update your Stripe API version, but that was only part of the answer in our case.

@pippinsplugins With Stripe in particular, sometimes the `charge.succeeded` event doesn't seem to register at all. Not in Stripe's logs and not in RCP. This only seems to happen on renewals. As a result, the expiration dates doesn't get set and those customers get expired immediately. Any idea what might be going on here?

**Comment by pippinsplugins** [comment]
@kaizau Usually that means the customer's account had a balance on it. When there is a balance that covers the amount of the charge, a `charge.succeeded` event is not fired. Instead a `invoice.payment_succeeded` event is fired.

**Comment by kaizau** [comment]
Updated the plugin to 2.3.8 and this still seems to be happening.

You're right that it has to do with the balance. I've been testing this by manually expiring subscribers from the admin panel. Initially, this has no effect on Stripe's event log. But when I use the same subscriber to renew, Stripe's event log shows the following (in order of occurrence) :

```
A proration adjustment for -$38.99 was created for [user]
A proration adjustment for $194.97 was created for [user]
[user] upgraded to [plan].
```

It does not log a `invoice.payment_succeeded` event, and as soon as the user refreshes the page, his account is immediately marked expired, since the expiration date was not updated.

In other words, neither `charge.succeeded` nor `invoice.payment_succeeded` are fired, meaning the expiration date is never touched.

Thoughts?

**Comment by pippinsplugins** [comment]
Which events are fired when the adjustment is made?

**Comment by kaizau** [comment]
In this order:

* customer.updated (deletes old card)
* customer.source.created (adds new card)
* invoiceitem.created (negative proration adjustment)
* invoiceitem.created (positive proration adjustment)
* customer.subscription.updated (changes plan, in this case, a downgrade)

**Comment by pippinsplugins** [comment]
@kaizau Are you on the latest version of the Stripe API?

**Comment by kaizau** [comment]
Upgraded on 10/12. Doesn't look like there have been any major changes since? https://stripe.com/docs/upgrades

**Comment by pippinsplugins** [comment]
Related to #293

**Comment by pippinsplugins** [comment]
Related #299

**Comment by pippinsplugins** [comment]
@kaizau Your comments from the 30th indicate the source of the problem. Your customer never has an `invoice.payment_succeeded` or `charge.succeeded` event fire. Those are the only two events that trigger an expiration update.

I'm not yet sure how to best resolve that so I'm going to leave this open pending further investigation.

**Comment by kaizau** [comment]
Indeed. Question is why those events don't fire (or only fire intermittently).

I'm currently using a workaround that sets the renewal date immediately after a payment is initiated, but that's obviously not ideal.

**Comment by pippinsplugins** [comment]
That would be a question best answered by Stripe :)

**Comment by tannerm** [comment]
@kaizau I am working on fixing this issue and I have not been able to replicate it. Can you share the steps to recreate it?
